File: day20/Day20.hs
import Data.Function
import Data.Functor
import qualified Data.List as L
import qualified Data.Map.Strict as M
import Data.Maybe

import System.Environment

type Algorithm = M.Map Int Bool
type Image = M.Map (Int, Int) Bool
type Input = (Algorithm, Image)

parse :: [String] -> Input
parse (code : _ : img) =
  ( M.fromList . zip [0 ..] . map (== '#') $ code
  , M.fromList $ [((y, x), v == '#') | (y, row) <- zip [0 ..] img, (x, v) <- zip [0 ..] row]
  )

boolsToInt :: [Bool] -> Int
boolsToInt = L.foldl' (\a b -> a * 2 + if b then 1 else 0) 0

step :: Algorithm -> Int -> Image -> Image
step algo n img =
  let empty =
        if even n
          then algo M.! 0
          else algo M.! boolsToInt (replicate 9 (algo M.! 0))
      [minY, maxY, minX, maxX] =
        [(minimum, fst), (maximum, fst), (minimum, snd), (maximum, snd)] <&> \(f, g) ->
          f . map g . M.keys $ img
   in M.fromList
        [ ((y, x), nv)
        | y <- [minY - 1 .. maxY + 1]
        , x <- [minX - 1 .. maxX + 1]
        , let nv =
                (algo M.!) . boolsToInt $
                  [ fromMaybe empty $ M.lookup (y + dy, x + dx) img
                  | dy <- [-1 .. 1]
                  , dx <- [-1 .. 1]
                  ]
        ]

solveGen :: Input -> [Image]
solveGen (algorithm, img) = L.scanl' (&) img . map (step algorithm) $ [1 ..]

solve1 = length . filter id . M.elems . (!! 2) . solveGen
solve2 = length . filter id . M.elems . (!! 50) . solveGen

solutions =
  [ solve1
  , solve2
  ]

main :: IO ()
main = do
  idx <- read . head <$> getArgs
  print
    . (solutions !! (idx - 1))
    . parse
    . lines
    =<< getContents

File: day20/Day20.py
from sys import argv, stdin
from collections import defaultdict

text = stdin.read()
lines = text.splitlines()

algo = lines[0]
img = defaultdict(
    lambda: ".", {(y, x): v for y, l in enumerate(lines[2:]) for x, v in enumerate(l)}
)


def get_ranges(img):
    return (
        min(map(lambda x: x[0], img)),
        min(map(lambda x: x[1], img)),
        max(map(lambda x: x[0], img)),
        max(map(lambda x: x[1], img)),
    )


count = 0


def get_val(data):
    data = data.replace(".", "0")
    data = data.replace("#", "1")
    data = int(data, 2)
    data = algo[data]
    return data


def step(img):
    global count
    count += 1
    min_y, min_x, max_y, max_x = get_ranges(img)

    default = algo[0] if count % 2 == 1 else get_val(algo[0] * 9)
    new_img = defaultdict(lambda: default)
    for y in range(min_y - 1, max_y + 2):
        for x in range(min_x - 1, max_x + 2):
            data = "".join(
                img[(y + dy, x + dx)] for dy in range(-1, 2) for dx in range(-1, 2)
            )
            data = get_val(data)

            new_img[(y, x)] = data

    return new_img


if argv[1] == "1":
    img = step(step(img))
else:
    for i in range(50):
        img = step(img)

print(sum(v == "#" for v in img.values()))
